Core/Src/main.c

```diff
@@ -86,18 +113,60 @@ int main(void)
 
   /* Initialize all configured peripherals */
   MX_GPIO_Init();
+  MX_DMA_Init();
   MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
   /* USER CODE BEGIN 2 */
 
+
+
+  // Enable Chip Select on default
+  HAL_GPIOWritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+
+
+  // Serial terminal test
+  uart_buffer_length = sprintf(uart_buffer, "Serial terminal test\r\n");
+  HAL_UART_Transmit(&huart2, (uint8_t *) uart_buffer, uart_buffer_length, 100);
+
+
+  //Start the timer for PWM
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
+
+
+
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
   while (1)
   {
+
+	// One SPI communication cycle with MCP3004 ADC
+	HAL_GPIOWritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);
+	HAL_SPI_TransmitRecieve(&hspi1, spi_output_buffer, spi_input_buffer, 3, 100);
+	HAL_GPIOWritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+
+	//Extract 10 bit data in spi_input_buffer to spi_input_value
+	spi_input_value = (((uint16_t)(spi_input_buffer[1] & 0x03)) << 8)
+			          | (uint16_t) spi_input_buffer[2];
+
+	//set the PWM compare register value for 5-10% duty cycle depending on MSP3004 ADC
+	//  register value for 5% of PWM_COUNTER_COUNTS set when spi_input_value = 0
+	//  register value increases for up to an additional 5% when spi_input_value > 0, until MAX_SPI_INPUT_VALUE is reached.
+	//	additional '-1' since "period - 1" = counts
+	pwm_compare_register_counts = PWM_COUNTER_COUNTS*(1 + spi_input_value/MAX_SPI_INPUT_VALUE)*5/100 - 1;
+
```

Comment:
Just noticed this, `pwm_compare_register_counts` is a uint16_t type but it should be of type double. So in your calculation you have a 5/100 which I assume is for multiplying by 5% but when working with integers, 5/100 = 0 because integer division cuts off the decimal point.

So you should change the type to double and explicitly do floating point division by using 0.05 or  5.0/100.0

Reply:
Oops, I confused C with Python